### ISSUE TITLE
Stuck at activating dart DevTools

### ISSUE THREAD

**Issue by pot-code** [flutter/flutter-intellij#5308]
## Steps to Reproduce

Hit the debug button and it always shows the `starting devtools` popup (each time it gets restarted), which may seem like the duplication of #5234, but I think it's the desired behavior?

After started, the simulator shows nothing, and the console output (no `Debug service listening` nor `Syncing files to device...`:

```
Launching lib/main.dart on iPhone 12 Pro Max in debug mode...
Running Xcode build...
Xcode build done.    
```

Here comes the most confusing part, after some seconds or minutes, it suddenly prints an `Activating Dart DevTools...` message in the debug console, then it will get stuck.

After some digging, I find that the plugin has executed the command `pub.dart.snapshot global activate devtools` more than once, then I killed the extra process and it finally works. 

so my question is, is it normal to have the `starting devtools` popup each time I restart it? and why does it try to activate the Devtool even when I have activated it globally?

Thanks!

## Version info

```
[✓] Flutter (Channel stable, 2.0.1, on Mac OS X 10.15.7 19H524 darwin-x64)
    • Flutter version 2.0.1 at /Users/censored/flutter
    • Framework revision c5a4b4029c (5 days ago), 2021-03-04 09:47:48 -0800
    • Engine revision 40441def69
    • Dart version 2.12.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/censored/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Proxy Configuration
    • HTTP_PROXY is set
    • NO_PROXY is localhost,127.0.0.1
    • NO_PROXY contains 127.0.0.1
    • NO_PROXY contains localhost

[✓] Connected device (2 available)
    • iPhone 12 Pro Max (mobile) • A5DFC6FE-533D-4F75-A6DF-40A546141339 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-4
      (simulator)
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 88.0.4324.192

• No issues found!
```

**Comment by kenzieschmoll** [comment]
@helin24 Are we checking to see if DevTools is activated before we try to activate it? See [this method](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/devtools_launcher.dart#L156) in flutter_tools for an example of only attempting to activate DevTools if it is not active or if it has been more than N hours since the last time we tried (we may already be doing this in IntelliJ - I can't remember).

**Comment by kenzieschmoll** [comment]
@helin24 maybe we should be killing the devtools process if a user clicks Stop before the process to serve devtools has completed

**Comment by helin24** [comment]
It sounds like there are a few things going on here:
1. Is this on the latest version of the Flutter plugin? In v54 we skip waiting for DevTools to start on pressing debug after the first time a project is run. This doesn't fix whether DevTools gets started/canceled during the first attempt though.
2. It sounds like you also need a proxy to download DevTools properly - is this something you've set up, e.g. using these [instructions](https://flutter.dev/community/china), if you're working in China? And are able to use DevTools, e.g. if you start an app from the command line or after you've started successfully running an app in IntelliJ?
3. Can you describe how you stopped the debug process early? IntelliJ should be locked while the "Starting DevTools.." message is up.

**Comment by pot-code** [comment]
@helin24 thank you for your reply! as for the question:


1. sure, it's the latest version, flutter v54, dart v201.9335

**Comment by helin24** [comment]
Hmm.. this is strange. I can't reproduce being able to stop the debug process when the "Starting DevTools..." popup is shown. Can you post a screen capture of this, or a screenshot indicating what buttons you're using? I'd like to see if you're using something different than what I'm thinking of.

And then the "Activating DevTools" message is shown in the debug console after your app has started?

@kenzieschmoll for this version of flutter we're using the daemon method `devtools.serve` to start DevTools, so it looks like we should be going through the code you linked. Not sure why there would be multiple pub activations in this case.

**Comment by pot-code** [comment]
@helin24 well...I think there is a misunderstanding, maybe the Devtools should be started after the "Starting DevTools..." popup disappears, BUT...mine would disappear even if the Devtools not get installed or started, that's the point.

then I simply clicked the red square button to stop the debug...

so it prints the "Activating DevTools" message after the "Starting DevTools..." disappears, with random delay.

>Not sure why there would be multiple pub activations in this case.

Hmmm, I guess it's because it doesn't kill the downloading process, each time I stopped the debug, it will stack up.

**Comment by helin24** [comment]
Ah I see, this makes more sense. There's a timeout on "Starting DevTools.." so that we don't permanently block starting the app if DevTools startup is very slow or fails. If startup is slow, we let it continue in the background since we still want to have an instance of DevTools for your project. i think what could be happening is DevTools fails to start for your project, and then we try to start DevTools from the app again since we were unable to send a DevTools URL to the app. 

So does this happen consistently each time you open a project? i.e. The "Starting DevTools" popup appears the first time you run/debug your app, times out, and then you get the many pub processes? Or does this happen intermittently and get fixed once you kill the pub processes?

**Comment by pot-code** [comment]
@helin24 sorry for the late...if the Devtool doesn't get installed, it will time out, every time, and the pub process would get stacked up. But once you have killed these processes, **and be sure to install the Devtool before doing anything**, everything will work fine, no matter if you restart the IDE or restart the PC.

I just did a reinstallation to the Flutter but failed to reproduce the issue, the good thing is that while the Devtool is already get installed (I forgot to clean the .pub-cache folder), the debug process started successfully without any problems, which means the issue is gone.

so, to fix the issue, try killing all the pub processes and setting up the proxy if necessary, then you are ready to go only after you have installed the Devtools somewhere else.

**Comment by helin24** [comment]
This DevTools opening process has been changed pretty significantly since this issue was last discussed. I'm going to close this issue since problems with this using the latest plugin with stable Flutter will require a different discussion.

**Comment by helin24** [comment]
@OnClickListener2048 could you open a new issue with a description of what you're encountering? The code for this functionality has changed a lot since this bug was opened, so we'll need new information to debug.

**Comment by OnClickListener2048** [comment]
> [@OnClickListener2048](https://github.com/OnClickListener2048) could you open a new issue with a description of what you're encountering? The code for this functionality has changed a lot since this bug was opened, so we'll need new information to debug.

will do.
